### PR TITLE
Removes carriers.

### LINKED ISSUE #37: Refactor inject/extract APIs
From @lvht in https://github.com/opentracing/opentracing-php/issues/31

Please all me to make another comment on the following API,
> function inject(SpanContext $spanContext, $format, Writer $carrier)
> function extract($format, Reader $carrier)

Both the Writer and Reader interface is needless, because it should defined in different implementations.

I think we should change them to
> function inject(SpanContext $spanContext, $format, &$carrier)
> function extract($format, $carrier)

Maybe we could make the `inject` method return a $carrier instance.

So that we can use any type to transfer the span context.

With the old API, we should
```php
<?php
$tracer = $factory->initTracer('foo');

$data = ['trace-id' => $_SERVER['HTTP_UBER_TRACE_ID']];
$textMap = TextMap::fromArray($data);
$context = $tracer->extract('text_map', $textMap);
$span = $tracer->startSpan('bar', ['child_of' => $context]);


$textMap = TextMap::fromArray([]);
$tracer->inject($clientSapn->getContext(), 'text_map', $textMap);
$injectTarget = $textMap->getIterator()->getArrayCopy();
```
But with the API I proposed, we only
```php
<?php
$tracer = $factory->initTracer('foo');

$trace_id = $_SERVER['HTTP_UBER_TRACE_ID'];
$context = $tracer->extract('text_map', $trace_id);
// even we could
$context = $tracer->extract('text_map', $_SERVER['HTTP_UBER_TRACE_ID']);
$span = $tracer->startSpan('bar', ['child_of' => $context]);

$tracer->inject($clientSapn->getContext(), 'text_map', $trace_id);
header("Trace-Id: $trace_id");
```

All in all, we should not restrict the type of carrier. Remember that we are using the language of PHP, a weak type, dynamic language, not Java, not Go, not Python.

### PR DESCRIPTION
**Pull request by jcchavezs**
This PR removes carriers (and interfaces Reader and Writer) as suggested by @lvht.

Reasons are:
1. By providing carriers we also dictate how injectors/extractors should work and for that we add a lot of constraints and complexity around them that they might not need. In addition, there is no specific structure for baggage items so forcing them to use an strict carrier will add some overhead we want to avoid.
2. There is a lot of noise around how PHP frameworks work with headers so the arguments of providing a standard carrier that will work in most of the cases is invalid. For example, Symfony request lowercase the header keys so a custom extractor should be written for B3 extraction.

Ping @felixfbecker @beberlei @lvht @tedsuo @yurishkuro @tedsuo 

Closes #37

### CODE REVIEW

**Comment by lvht** [comment]
:+1:

File: README.md [hunk 1]
```diff
@@ -105,17 +106,16 @@ reference.
...
     
     ...
     
     $tracer = GlobalTracer::get(); 
     
     $spanContext = $tracer->extract(
-        Propagator::HTTP_HEADERS,
-        HttpHeaders::withHeaders($request->getHeaders())
+        Formats\HTTP_HEADERS,
+        $request
```
Comment (yurishkuro) [inline README.md]: how is the tracer implementation supposed to know how to operate on the `request`? Different frameworks have different notions of the request, the tracer cannot be responsible for knowing how to access each of them. The Carrier APIs like TextMapReader provide the necessary level of abstraction by giving the tracer a stable API to work against, and the carrier translates that API into the operations on the underlying `request` object (or more typically, request headers).

Comment (jcchavezs) [reply README.md]: Fair point @yurishkuro I just updated the documentation, please have a look.

Regarding the TextMapReader I don't think we need that anymore. The `array` and the `ArrayObject` types are enough for this purposes.

**Comment by yurishkuro** [comment]
@jcchavezs do you mean that you're using standard types as carriers? E.g. in Python the carrier for the TextMap format is the default dictionary. If so, I think that's the right approach, but it doesn't seem to be mentioned in the documentation for the formats - see Go API that clearly states which carrier is expected with which format.

On a separate point, I would suggest implementing a Mock tracer (#40) so that changes like this can be validated in an actual implementation, rather than just the API.

File: src/OpenTracing/Formats.php [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+ * Unlike `TEXT_MAP`, the `HTTP_HEADERS` format requires that the keys and values be valid as HTTP headers as-is
+ * (i.e., character casing may be unstable and special characters are disallowed in keys, values should be
+ * URL-escaped, etc).
+ *
+ * For both Tracer::inject() and Tracer::extract() the carrier must be a `array|ArrayObject`.
+ *
+ * For example, Tracer::inject():
+ *
+ *    $headers = $request->getHeaders()
+ *    $tracer->inject($span->getContext(), Formats\HTTP_HEADERS, $headers)
```
Comment (yurishkuro) [inline src/OpenTracing/Formats.php]: I think the examples mixed up inject and extract.

File: src/OpenTracing/Formats.php [hunk 3]
```diff
@@ -0,0 +1,45 @@
+<?php
+
+namespace OpenTracing\Formats;
+
+/**
+ * Used a (single) arbitrary binary blob representing a SpanContext
```
Comment (yurishkuro) [inline src/OpenTracing/Formats.php]: what is the format of the carrier for binary?

File: src/OpenTracing/Formats.php [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+ */
+const BINARY = 'binary';
+
+/**
+ * Used for an arbitrary string-to-string map with an unrestricted character set for both keys and values
+ *
+ * Unlike `HTTP_HEADERS`, the `TEXT_MAP` format does not restrict the key or
+ * value character sets in any way.
+ *
+ * For both Tracer::inject() and Tracer::extract() the carrier must be a `array|ArrayObject`.
```
Comment (yurishkuro) [inline src/OpenTracing/Formats.php]: just for my info, is array/ArrayObject in PHP an associative array (like in Javascript)?

Comment (felixfbecker) [reply src/OpenTracing/Formats.php]: Every array in PHP is associative, _unlike_ Javascript

Comment (yurishkuro) [reply src/OpenTracing/Formats.php]: Thanks. I meant "like in Javascript objects"

File: src/OpenTracing/Formats.php (hunk 3 above)
Comment (jcchavezs) [reply src/OpenTracing/Formats.php]: I am not entirely sure. I'd say it is a `resource` but honestly I don't have too much experience with it. Any suggestion @felixfbecker @beberlei?

Comment (felixfbecker) [reply src/OpenTracing/Formats.php]: It's probably just `string`. Strings in PHP are binary (not UTF8 like in JS) and are used for binary content. `resource` is for things like a file handle, a stream, a socket connection etc.

**Comment by tedsuo** [comment]
This looks good to me. One question: for http headers, is []string more useful that [][]string? I don't think tracers will inject/extract more than one value per header. But if most http header implementations use [][]string perhaps we should match that? I believe I saw this discussed before, and []string is fine. Just double checking.

**Comment by felixfbecker** [comment]
Can this be merged?

**Comment by lvht** [comment]
I wish this PR to be merged as early as possible.

**Comment by lvht** [comment]
@jcchavezs Would you like make a v0.x.x tag? I need to require a release, even it is unstable.

**Comment by felixfbecker** [comment]
The next version is 1.0.0-beta2

**Comment by lvht** [comment]
ping @jcchavezs

**Comment by jcchavezs** [comment]
https://github.com/opentracing/opentracing-php/tree/1.0.0-beta2

**Comment by jcchavezs** [comment]
@tedsuo so most implementations use `string[][]`, problem is that there are different ways to get headers, for example [getallheaders](http://php.net/manual/en/function.getallheaders.php) will return `string[]` having duplicated values glued by a comma whereas the PSR7 implementation returns a `string[][]`. I believe every vendor would setup something that works for them based on their propagation model.

**Pull request merged**